allocator.go

```diff
@@ -168,39 +171,65 @@ func (a *allocator) adhereToChurnConstraint() {
 
 // allocate is a terminal method call that returns a status and paired allocation.
 // The status could be false if the existing model is invalid or unsatisfiable.
-func (a *allocator) allocate() (ok bool, allocation Allocation) {
+func (a *allocator) allocate() (allocation Allocation, err error) {
 
 	// iterate over replicas, assign each replicaID a list of IV's sized r.rf.
 	// These will ultimately then read as: replicaID's replicas assigned to nodes [N.1, N.2,...N.RF]
 	for _, r := range a.replicas {
+		if r.rf > len(a.nodes) {
+			return nil, RfGreaterThanClusterSizeError(r.id)
+		}
 		a.assignment[r.id] = make([]solver.IntVar, r.rf)
 		for j := range a.assignment[r.id] {
 			// constrain our IV's to live between [0, len(nodes) - 1].
 			a.assignment[r.id][j] = a.model.NewIntVarFromDomain(
 				solver.NewDomain(int64(a.nodes[0].id), int64(a.nodes[nodeId(len(a.nodes)-1)].id)),
 				fmt.Sprintf("Allocation var for r.id:%d.", r.id))
 		}
+		a.model.AddConstraints(solver.NewAllDifferentConstraint(a.assignment[r.id]...))
 	}
 	// add constraints given opts/configurations.
 	if a.opts.withResources {
+		totalDemand := int64(0)
+		totalResource := int64(0)
+		for _, re := range []Resource{DiskResource} {
+			for _, r := range a.replicas {
+				totalDemand += r.demands[re]
+			}
+			for _, n := range a.nodes {
+				totalResource += n.resources[re]
+			}
+		}
+		if totalDemand > totalResource {
+			return nil, InsufficientClusterCapacityError{}
+		}
 		a.adhereToResourcesAndBalance()
 	}
 
 	if a.opts.withTagAffinity {
+		for _, r := range a.replicas {
+			foundNode := false
+			for _, n := range a.nodes {
+				if replicaTagsAreSubsetOfNodeTags(r.tags, n.tags) {
+					foundNode = true
+					break
+				}
+			}
+			if !foundNode {
+				return nil, RangeWithWaywardTagsError(r.id)
+			}
+		}
```

Comment:
Might consider caching these, pre-mature optimizations are bad but caching could actually simplify some of the code in the `adhereToNodeTags()` method.